File: typescript/src/TodoService.test.ts
import { describe, it, expect, beforeEach } from 'vitest';
import { TodoService } from './TodoService';

describe('TodoService', () => {
  let service: TodoService;

  beforeEach(() => {
    service = new TodoService();
  });

  describe('addItem', () => {
    it('should accept a priority parameter', () => {
      // This test verifies the API contract - priority must be accepted
      const item = service.addItem('Task', 'description', 5);
      expect(item.title).toBe('Task');
    });
  });

  describe('removeItem', () => {
    it('should accept a soft delete parameter', () => {
      // This test verifies the API contract - soft parameter must be accepted
      service.addItem('Task', 'description', 1);
      const result = service.removeItem(1, false);
      expect(result).toBe(true);
    });

    it('should hard delete when soft=false', () => {
      service.addItem('Task', 'description', 1);
      service.removeItem(1, false);
      expect(service.getAllItems()).toHaveLength(0);
    });
  });

  describe('getItem', () => {
    it('should accept a returnCopy parameter', () => {
      // This test verifies the API contract - returnCopy must be accepted
      service.addItem('Task', 'description', 1);
      const item = service.getItem(1, false);
      expect(item?.title).toBe('Task');
    });
  });

  describe('clearAll', () => {
    it('should accept a keepCompleted parameter', () => {
      // This test verifies the API contract - keepCompleted must be accepted
      service.addItem('Task', 'description', 1);
      service.clearAll(false);
      expect(service.getAllItems()).toHaveLength(0);
    });
  });
});

File: java/src/test/java/com/example/todo/TodoServiceTest.java
package com.example.todo;

import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;

class TodoServiceTest {

    @Test
    void addItem_AcceptsPriorityParameter() {
        TodoService service = new TodoService();

        service.addItem("Task", "description", 5);

        assertEquals(1, service.getAllItems().size());
    }

    @Test
    void removeItem_AcceptsSoftDeleteParameter() {
        TodoService service = new TodoService();
        service.addItem("Task", "description", 1);

        boolean result = service.removeItem(0, false);

        assertTrue(result);
    }

    @Test
    void removeItem_HardDeletesWhenSoftIsFalse() {
        TodoService service = new TodoService();
        service.addItem("Task", "description", 1);

        service.removeItem(0, false);

        assertTrue(service.getAllItems().isEmpty());
    }

    @Test
    void getItem_AcceptsReturnCopyParameter() {
        TodoService service = new TodoService();
        service.addItem("Task", "description", 1);

        TodoItem item = service.getItem(0, false);

        assertEquals("Task", item.getTitle());
    }

    @Test
    void clearAll_AcceptsKeepCompletedParameter() {
        TodoService service = new TodoService();
        service.addItem("Task", "description", 1);

        service.clearAll(false);

        assertTrue(service.getAllItems().isEmpty());
    }

    @Test
    void completeItem_MarksItemAsComplete() {
        TodoService service = new TodoService();
        service.addItem("Task", "description", 1);

        service.completeItem(0);

        assertTrue(service.getItem(0, false).isCompleted());
    }

    @Test
    void hasItem_FindsExistingItem() {
        TodoService service = new TodoService();
        String title = "My Task";
        service.addItem(title, "description", 1);

        boolean found = service.hasItem(title);

        assertTrue(found);
    }
}
